### PR TITLE
[prop-types] Rework nominal type

### PR DESCRIPTION
**Pull request by milesj**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

to: @ferdaber I updated the nominal hack to an argument like the suggestion in the previous PR. This resolves a lot of issues, however, I'm not sure the best way to resolve the 2 failing lints.

### CODE REVIEW

File: types/prop-types/index.d.ts [hunk 1]
```diff
@@ -15,10 +15,14 @@ export type RequiredKeys<V> = { [K in keyof V]: V[K] extends Validator<infer T>
 export type OptionalKeys<V> = Exclude<keyof V, RequiredKeys<V>>;
 export type InferPropsInner<V> = { [K in keyof V]: InferType<V[K]>; };
 
-export interface Validator<T> {
-    (props: object, propName: string, componentName: string, location: string, propFullName: string): Error | null;
-    [nominalTypeHack]?: T;
-}
+export type Validator<T> = (
```
Comment (milesj) [inline types/prop-types/index.d.ts]: Linter errored when using an interface.

**Comment by ferdaber** [comment]
Looks good, could you re-enable the tests for the react types (`libraryManagedAttributes.tsx`) and check that they're all still kosher as well?

**Comment by milesj** [comment]
@ferdaber Enabled the tests. The work in 3.0+ but fail hard in < 2.9.

I've got another issue I'm hoping you have some direction on. I'm testing these react/prop-types changes in our codebase, and we also use `airbnb-prop-types`. The following code fails:

```js
ProgressCard.propTypes = {
  progress: childrenOfType(ProgressBar, SteppedProgressBar).isRequired,
};
```

With the error:

```
Type '{ progress: Validator<string | number | boolean | {} | ReactElement<any> | ReactNodeArray | ReactPortal>; }' is not assignable to type 'ValidationMap<ProgressCardProps>'.
  Types of property 'progress' are incompatible.
    Type 'Validator<string | number | boolean | {} | ReactElement<any> | ReactNodeArray | ReactPortal>' is not assignable to type 'Validator<ReactNode>'.
      Types of parameters '_nominalTypeParam' and '_nominalTypeParam' are incompatible.
        Type 'ReactNode' is not assignable to type 'string | number | boolean | {} | ReactElement<any> | ReactNodeArray | ReactPortal | undefined'.
          Type 'null' is not assignable to type 'string | number | boolean | {} | ReactElement<any> | ReactNodeArray | ReactPortal | undefined'.
```

But removing `isRequired` passes.

```js
ProgressCard.propTypes = {
  progress: childrenOfType(ProgressBar, SteppedProgressBar),
};
```

And this is our type definition:

```js
export function childrenOfType<T = React.ReactNode>(
  ...types: React.ReactType[]
): PropTypes.Requireable<T>;
```

Seems like more `null` shenanigans.

**Comment by ferdaber** [comment]
Those tests are supposed to fail < TS 3.0 (library managed attributes), so if they work in TS 3.0 we're good to go 👍

**Comment by ferdaber** [comment]
@milesj This is correct, IMO there is little education out there on how exactly the prop-types runtime works, but `NonNullable<ReactNode>` is the most sound type for `progress`, because if `null` or `undefined` is passed into `progress` it will throw a prop type warning.

**Comment by milesj** [comment]
@ferdaber Yeah. I went ahead and wrote an ESLint to check that `NonNullable` is used. Much easier to manage.

**Comment by stnwk** [comment]
What's the status here?

**Comment by milesj** [comment]
@stnwk There's 2 prop-type failures I'm not sure how to resolve.

**Comment by ferdaber** [comment]
`UnmatchedPropKeys` and `UnmatchedPropKeys2` tests should be deleted. I did find that they keep failing. If those are the only two that are failing we should be good to go still 👍

**Comment by milesj** [comment]
Ok cool, I'll delete them.

**Comment by milesj** [comment]
@ferdaber Removed. As for the react errors, looks like there are 2 that might be warranted and require a fix? Specifically lines 62 and 109.

**Comment by ferdaber** [comment]
Are you saying they are no longer throwing errors? Inspecting the file it looks like those lines have an `// $ExpectError` directive on them.

**Comment by milesj** [comment]
@ferdaber They do, but the linter/testing throws `ERROR: 109:1   expect  Expected an error on this line, but found none.`.

**Comment by ferdaber** [comment]
That's worrying... looks like it's breaking unannotated prop types' inference for library managed attributes. Specifically `.isRequired` is no longer causing that specific prop to be required.

**Comment by ferdaber** [comment]
What is the result of `InferType<typeof PropTypes.string.isRequired>`? It should be `string`.

**Comment by milesj** [comment]
Whoops :P

**Pull request closed without merging**